src/mlpack/core/optimizers/fw/update_linesearch.hpp

```diff
@@ -0,0 +1,194 @@
+ *
+ */
+class UpdateLineSearch
+{
+ public:
+  /**
+   * Construct the line search update rule.
+   *
+   * @param maxIter Max number of iterations in line search.
+   * @param tolerance Tolerance for termination of line search.
+   */
+  UpdateLineSearch(const size_t maxIter = 100000,
+                   const double tolerance = 1e-5) :
+      maxIter(maxIter), tolerance(tolerance)
+  {/* Do nothing */}
+  
+
+  /**
+   * Update rule for FrankWolfe, optimize with line search using secant method.
+   *
+   * FunctionType template parameters are required.
+   * This class must implement the following functions:
+   *
+   * FunctionType:
+   *
+   *   double Evaluate(const arma::mat& coordinates);
+   *   Evaluation of the function at specific coordinates.
+   *
+   *   void Gradient(const arma::mat& coordinates,
+   *                 arma::mat& gradient);
+   *   Solve the gradient of the function at specific coordinate,
+   *   returned in gradient.
+   *
+   * @param function function to be optimized,
+   * @param oldCoords previous solution coordinates, one end of line search.
+   * @param s current linear_constr_solution result, the other end point of
+   *        line search.
+   * @param newCoords output new solution coords.
+   * @param numIter current iteration number, not used here.
+   */
+  template<typename FunctionType>
+  void Update(FunctionType& function,
+              const arma::mat& oldCoords,
+              const arma::mat& s,
+              arma::mat& newCoords,
+              const size_t numIter)
+
+  {
+    double gamma = LineSearchSecant(function, oldCoords, s);
+    newCoords = (1 - gamma) * oldCoords + gamma * s;
+  }
+
+  //! Get the tolerance for termination.
+  double Tolerance() const {return tolerance;}
+  //! Modify the tolerance for termination.
+  double& Tolerance() {return tolerance;}
+  
+  //! Get the maximum number of iterations (0 indicates no limit).
+  size_t MaxIterations() const { return maxIter; }
+  //! Modify the maximum number of iterations (0 indicates no limit).
+  size_t& MaxIterations() { return maxIter; }
+
+ private:
+  //! Tolerance for convergence.
+  double tolerance;
+  
+  //! Max number of iterations.
+  size_t maxIter;
+
+  /**
+   * Line search to minimize function between two points with Secant method,
+   * that is, to find the zero of Derivative(gamma), where gamma is in [0,1].
+   *
+   * The function is assumed to be convex here, otherwise might not converge.
+   *
+   * If the function is convex, Derivative(gamma) would be a nondecreasing
+   * function of gamma, then the soluton exists.
+   * If the function is strongly convex, Derivative(gamma) is strictly
+   * increasing, then the solution exists and is unique.
+   *
+   * @param function function to be minimized.
+   * @param x1 One end point.
+   * @param x2 The other end point.
+   * @return Optimal solution position ratio betwen two points,
+   *         0 means x1, 1 means x2.
+   */
+  template<typename FunctionType>
```

Comment:
As we discussed earlier, consider moving this to it's own file in `core/optimizers/line_search`.